### ISSUE TITLE
Alertmanager not taking multiple secrets with multiple yaml files

### ISSUE THREAD

**Issue by yaminisridaran** [prometheus-operator/prometheus-operator#2258]
**What did you do?**
Alertmanager Pods are up and running with Alertmanager-example secret which has Alertmanager.yaml (different set of receivers)
Added another secret Alertmanager-example1 to the same namespace which has Alertmanager1.yaml (different set of receivers)
**What did you expect to see?**
I have multiple users need to configure their receivers in multiple secrets with multiple yaml. Will Alertmanager consider multiple secrets with different yaml files?

**What did you see instead? Under which circumstances?**
Alertmanager Pods should considers the secret Alertmanager-example and loaded Alertmanager.yaml
Alertmanager Pods is not considering Alertmanager-example1 and not loading the alertmanager1.yaml

**Environment**
k8s

**Comment by brancz** [comment]
This is a feature we've been thinking about for a while as well. We haven't come up with an API that we liked, but open for the discussion around it! :slightly_smiling_face:

**Comment by yaminisridaran** [comment]
Thank You for the response. Sure we can discuss. We have multiple applications having their particular receivers. It would be Great if have a crd or api to combine all the configuration file to a single configuration and provide to Alertmanager.

**Comment by yaminisridaran** [comment]
@brancz I am open for a discussion regarding this feature. Please let me know for the discussion.
I think we need this feature it would be great.

**Comment by yaminisridaran** [comment]
@brancz I am creating a document regarding this design. Shall I upload here or I need to send it to a mail list? Can you please guide me.

**Comment by brancz** [comment]
Sharing it here would be best I think, so we keep all communication in one place. Thanks a ton for looking into this!

**Comment by abursavich** [comment]
I would like to suggest tackling https://github.com/coreos/prometheus-operator/issues/2213 along with this. As it is now, you must embed API keys directly in the Alertmanager config. It would be nice to have those handled separately and injected into the rendered config by Secret reference. This would allow users to check their unencrypted Alertmanager config(s) into source control and manage their keys separately.

**Comment by yaminisridaran** [comment]
is there a way to upload the design overview doc here ?

**Comment by brancz** [comment]
You can create a pull request into the “/docs/design” directory adding your proposal. Then it’s easier to review :)

**Comment by yaminisridaran** [comment]
@brancz I think I am missing the location /docs/design directory.

**Comment by brancz** [comment]
it doesn't exist yet, but feel free to create it! :)

**Comment by madushan1000** [comment]
Came across with this project(https://github.com/dbsystel/alertmanager-config-controller), wondering if we could utilize this with a couple of new crd types to provide alert rule live reloading.

**Comment by brancz** [comment]
Generally I'd be interested in integrating something if anyone has figured this out. At first glance I see a potential issue with the linked controller's strategy, as it throws things together globally, whereas Alertmanager configuration is order sensitive, which can have various side effects.

**Comment by madushan1000** [comment]
I'm not sure if this is a proper way of dealing with something like this, but what if we modify the controller to order the alert rules according to a manually defined priority value(integer) in the crd/configmap objects?

**Comment by Kirchen99** [comment]
We always have `continue: true` at the first level of routes. So the order of the configuration is not so sensitive for us.
```route:
  receiver: default
  group_by:
  - alertname
  - instance
  routes:
  - receiver: team1
    match_re:
      receiver: ^(?:.*team1.*)$
    continue: true
  - receiver: team2
    match_re:
      receiver: ^(?:(.*,)*team2(,.*)*)$
    continue: true
  - receiver: team3
    match_re:
      receiver: ^(?:(.*,)*team3(,.*)*)$
    continue: true
    routes:
    - receiver: team3a
      match_re:
        label: ^(?:.*value.*)$
```

**Comment by brancz** [comment]
Right, that's exactly what I was thinking, is this automated in the controller?

**Comment by Kirchen99** [comment]
> Right, that's exactly what I was thinking, is this automated in the controller?

Not yet, We will implement this in the next release

**Comment by brancz** [comment]
Nice, I like that a lot. I think that and a few other things (like receiver separation) are pretty much what I was imagining. I actually also quite like your approach of using plain secrets/configmaps for this. Would you be interested in contributing this feature to the prometheus operator?

**Comment by mitchellmaler** [comment]
Is there any update on this? Just ran into the need to dynamically add notifiers outside of the main prom stack deployment. Either multiple secret or CRD support would be amazing! Is there anything that can help speed this support along?

**Comment by abursavich** [comment]
@mitchellmaler, check out #3451

**Comment by mitchellmaler** [comment]
@abursavich Thanks!

**Comment by paulfantom** [comment]
This is implemented with #3451
